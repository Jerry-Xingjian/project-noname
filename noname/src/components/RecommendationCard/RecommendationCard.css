.recommendation-card {
    width: 360px;
    border-radius: 10px;
    border: 1px solid #e5e5e5;
}

.recommendation-card .card-text small{
    max-width: 100px;
}