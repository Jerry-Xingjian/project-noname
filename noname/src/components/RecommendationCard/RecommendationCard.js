/* eslint-disable no-underscore-dangle */
import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import './RecommendationCard.css';
import { follow, unfollow } from '../../utils/api/follow';

function RecommendationCard(props) {
  // TODO: `Follow` button on click event, Call API
  const { recommendedInfo, userInfo } = props;
  const [isFollowed, setIsFollowed] = useState(false);

  const handleFollow = async () => {
    await follow(userInfo._id, recommendedInfo._id);
    // TODO: Update Recommendation
    setIsFollowed(true);
    // TODO: Call up an alert `successfully followed`
  };

  const handleUnfollow = async () => {
    // Remove following
    await unfollow(userInfo._id, recommendedInfo._id);
    // TODO: Update recommendations
    // await updateRecommendations(currentUser.current);
    setIsFollowed(false);
    // TODO: Pop up a confirm box
  };

  return (
    <div className="recommendation-card d-flex py-1 px-1 my-1">
      <img
        className="avatar-image ms-3 my-2 mx-2"
        src={recommendedInfo.profilePicture}
        alt="Avatar"
        mode="aspectFit"
      />
      <div className="card-body mx-2" id="direct-profile-page">
        <Link to={`/profile/${recommendedInfo._id}`} className="text-decoration-none" role="presentation">
          <h6 className="card-title mt-1" id="username-id">{ recommendedInfo.username }</h6>
          <div className="card-text"><small className="text-muted d-inline-block text-truncate">{ recommendedInfo.bio }</small></div>
        </Link>
      </div>
      {
        isFollowed ? (
          <button type="button" id="unfollow-btn" className="btn btn-link text-decoration-none text-secondary" onClick={handleUnfollow}>Unfollow</button>
        ) : (
          <button type="button" id="follow-btn" className="btn btn-link text-decoration-none" onClick={handleFollow}>Follow</button>
        )
      }
    </div>
  );
}

export default RecommendationCard;
