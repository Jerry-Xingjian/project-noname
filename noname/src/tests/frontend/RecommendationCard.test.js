import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from '@testing-library/user-event';
import React from "react";
import '@testing-library/jest-dom'
import RecommendationCard from '../../components/RecommendationCard/RecommendationCard';
import {BrowserRouter, Router} from 'react-router-dom';
import renderer from 'react-test-renderer';
import {createMemoryHistory} from 'history';

const prop = {
  recommendedInfo: {
		"_id": "6388320305d016d46329331d",
		"email": "qhao@seas.upenn.edu",
		"password": "$2b$10$Mxx3KERe2hI9GUEkb8Hsm.jr2gLNJyZYfooiGYK1RW0QBPdComYSO",
		"username": "qhao",
		"profilePicture": "https://nonome-project-media.s3.amazonaws.com/default_avatar.png",
		"followings": [],
		"followers": [],
		"createdAt": "2022-12-01T04:48:03.298Z",
		"updateAt": "2022-12-01T04:48:03.298Z",
		"bio": "This is some bio of the user qhao.\nThis can be super long but it's fine."
  },
	userInfo: {
		"_id": "6388320305d016d46329331d",
		"email": "qhao@seas.upenn.edu",
		"password": "$2b$10$Mxx3KERe2hI9GUEkb8Hsm.jr2gLNJyZYfooiGYK1RW0QBPdComYSO",
		"username": "qhao",
		"profilePicture": "https://nonome-project-media.s3.amazonaws.com/default_avatar.png",
		"followings": [],
		"followers": [],
		"createdAt": "2022-12-01T04:48:03.298Z",
		"updateAt": "2022-12-01T04:48:03.298Z",
		"bio": "This is some bio of the user qhao.\nThis can be super long but it's fine."
	}
}

describe("snapshot test", () => {
  test('RecommendationCard matches snapshot', () => {
    const component = renderer.create(
      <BrowserRouter>
          <RecommendationCard {...prop}/>
      </BrowserRouter>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe("Element existence test", () => {
  test('renders Follow text', () => {
    render(
      <BrowserRouter>
          <RecommendationCard {...prop}/>
      </BrowserRouter>
    );
    const followText = screen.getByText('Follow');
    expect(followText).toBeInTheDocument();
    expect(followText.innerHTML).toBe('Follow');
  });
})

describe("Event test", () => {
  test('click card title and navigate to other profile page', async () => {
    const history = createMemoryHistory();
    render(
      <Router location={history.location} navigator={history}>
          <RecommendationCard {...prop}/>
      </Router>
    );
    const user = userEvent.setup()
    user.click(screen.getByRole("presentation"))
    expect(history.location.pathname).toBe('/');
  }) 
})

describe("Click Event test", () => {
  test('click follow', async () => {
    render(
      <BrowserRouter >
          <RecommendationCard {...prop}/>
      </BrowserRouter>
    );
    const follow = screen.getByRole('button', { name: 'Follow' })
    await waitFor(() => {
      userEvent.click(follow)
    });
  });
})