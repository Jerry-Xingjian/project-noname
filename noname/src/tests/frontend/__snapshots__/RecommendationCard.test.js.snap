// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test RecommendationCard matches snapshot 1`] = `
<div
  className="recommendation-card d-flex py-1 px-1 my-1"
>
  <img
    alt="Avatar"
    className="avatar-image ms-3 my-2 mx-2"
    mode="aspectFit"
    src="https://nonome-project-media.s3.amazonaws.com/default_avatar.png"
  />
  <div
    className="card-body mx-2"
  >
    <a
      className="text-decoration-none"
      href="/profile/6388320305d016d46329331d"
      onClick={[Function]}
      role="presentation"
    >
      <h6
        className="card-title mt-1"
      >
        qhao
      </h6>
      <div
        className="card-text"
      >
        <small
          className="text-muted d-inline-block text-truncate"
        >
          This is some bio of the user qhao.
This can be super long but it's fine.
        </small>
      </div>
    </a>
  </div>
  <button
    className="btn btn-link text-decoration-none"
    onClick={[Function]}
    type="button"
  >
    Follow
  </button>
</div>
`;
